# 4. In-Class Exercises
Each week, we will have in-class exercises that give you some pratice on the concepts learned in class.


## 4.1. Exercise 1 (5 points)
For this exercise, we are going to use the table from the earlier examples. Instead of reading the table from a CSV, let's read the table directly from the [Wikipedia page on Metropolitan Statistical Areas](https://en.wikipedia.org/wiki/Metropolitan_statistical_area).

In [1]:
import pandas as pd
msa_by_pop = pd.read_html("https://en.wikipedia.org/wiki/Metropolitan_statistical_area",encoding="utf-8",)[1]
## It seems "latin-1" format doesn't work for me in VS Code, so I replaced it to UTF-8.

msa_by_pop.columns =['Rank', 'MSA', 'population_2021_est', 'population_2020',
       'perc_change', 'Encompassing_combined_statistical_area']

In [2]:
#Take a look at the dataset
msa_by_pop.head()

Rank                                           MSA  population_2021_est  \
0     1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA             19768458   
1     2        Los Angeles-Long Beach-Anaheim, CA MSA             12997353   
2     3        Chicago-Naperville-Elgin, IL-IN-WI MSA              9509934   
3     4           Dallas-Fort Worth-Arlington, TX MSA              7759615   
4     5      Houston-The Woodlands-Sugar Land, TX MSA              7206841   

   population_2020 perc_change Encompassing_combined_statistical_area  
0         20140470      −1.85%       New York-Newark, NY-NJ-CT-PA CSA  
1         13200998      −1.54%         Los Angeles-Long Beach, CA CSA  
2          9618502      −1.13%       Chicago-Naperville, IL-IN-WI CSA  
3          7637387      +1.60%           Dallas-Fort Worth, TX-OK CSA  
4          7122240      +1.19%          Houston-The Woodlands, TX CSA

Which is the MSA with the 10th largest population as estimated for 2021? 

In [3]:
## Insert your code here
msa_by_pop.sort_values('population_2021_est',ascending = False).iloc[[9]]

Rank                            MSA  population_2021_est  population_2020  \
9    10  Phoenix-Mesa-Chandler, AZ MSA              4946145          4845832   

  perc_change Encompassing_combined_statistical_area  
9      +2.07%                   Phoenix-Mesa, AZ CSA

Select the smallest 10 MSAs by 2020 Census population

In [4]:
## Insert your code here
msa_by_pop.sort_values('population_2020').head(10)

Rank                         MSA  population_2021_est  population_2020  \
383   384         Carson City, NV MSA                58993            58639   
381   382         Walla Walla, WA MSA                62682            62584   
382   383                Enid, OK MSA                61926            62846   
380   381         Lewiston, ID-WA MSA                64851            64375   
379   380            Danville, IL MSA                73095            74188   
378   379        Grand Island, NE MSA                76175            77038   
377   378              Casper, WY MSA                79555            79955   
375   376          Hinesville, GA MSA                82863            81424   
376   377            Columbus, IN MSA                82475            82208   
374   375  Bloomsburg-Berwick, PA MSA                82959            82863   

    perc_change  Encompassing_combined_statistical_area  
383      +0.60%        Reno-Carson City-Fernley, NV CSA  
381      +0.16%  Kennewick-Richland-Walla Walla, WA CSA  
382      −1.46%                                     NaN  
380      +0.74%                                     NaN  
379      −1.47%                                     NaN  
378      −1.12%                                     NaN  
377      −0.50%                                     NaN  
375      +1.77%  Savannah-Hinesville-Statesboro, GA CSA  
376      +0.32%      Indianapolis-Carmel-Muncie, IN CSA  
374      +0.12%      Bloomsburg-Berwick-Sunbury, PA CSA

Find the total 2020 Census population of these 384 MSAs. 

In [5]:
## Insert your code here
msa_by_pop['population_2020'].sum()

286104556

Based on the population growth between 2020 and 2021, let's estimate the population in 2021 using the following function: 

$$
pop_{2020} =pop_{2010}*(1+ {\% change}^t)
$$

where $t=10$ and $\% change$ is the percentage estimated change in population from 2020.

First, calculate the percentage change between 2020 and 2021 and create a new column with these values called `year_change`. Do not multiply by 100, just keep the values in their original decimal form.

In [6]:
## Insert your code here
msa_by_pop['year_change'] = msa_by_pop['population_2021_est'] / msa_by_pop['population_2020'] - 1
msa_by_pop

Rank                                           MSA  population_2021_est  \
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA             19768458   
1       2        Los Angeles-Long Beach-Anaheim, CA MSA             12997353   
2       3        Chicago-Naperville-Elgin, IL-IN-WI MSA              9509934   
3       4           Dallas-Fort Worth-Arlington, TX MSA              7759615   
4       5      Houston-The Woodlands-Sugar Land, TX MSA              7206841   
..    ...                                           ...                  ...   
379   380                              Danville, IL MSA                73095   
380   381                           Lewiston, ID-WA MSA                64851   
381   382                           Walla Walla, WA MSA                62682   
382   383                                  Enid, OK MSA                61926   
383   384                           Carson City, NV MSA                58993   

     population_2020 perc_change  Encompassing_combined_statistical_area  \
0           20140470      −1.85%        New York-Newark, NY-NJ-CT-PA CSA   
1           13200998      −1.54%          Los Angeles-Long Beach, CA CSA   
2            9618502      −1.13%        Chicago-Naperville, IL-IN-WI CSA   
3            7637387      +1.60%            Dallas-Fort Worth, TX-OK CSA   
4            7122240      +1.19%           Houston-The Woodlands, TX CSA   
..               ...         ...                                     ...   
379            74188      −1.47%                                     NaN   
380            64375      +0.74%                                     NaN   
381            62584      +0.16%  Kennewick-Richland-Walla Walla, WA CSA   
382            62846      −1.46%                                     NaN   
383            58639      +0.60%        Reno-Carson City-Fernley, NV CSA   

     year_change  
0      -0.018471  
1      -0.015426  
2      -0.011287  
3       0.016004  
4       0.011878  
..           ...  
379    -0.014733  
380     0.007394  
381     0.001566  
382    -0.014639  
383     0.006037  

[384 rows x 7 columns]

Does `year_change` line up with the `% change` column that we already have? Select the two columns together to do an eyeball estimate. 

In [7]:
## Insert your code here
msa_by_pop[['perc_change','year_change']]

perc_change  year_change
0        −1.85%    -0.018471
1        −1.54%    -0.015426
2        −1.13%    -0.011287
3        +1.60%     0.016004
4        +1.19%     0.011878
..          ...          ...
379      −1.47%    -0.014733
380      +0.74%     0.007394
381      +0.16%     0.001566
382      −1.46%    -0.014639
383      +0.60%     0.006037

[384 rows x 2 columns]

In [8]:
"""
Answer to the question: The numbers are the same, but the data types are not. Since the 'perc_change' column contains symbol "%" "+" and "-",
it is taken as objects rather than float numbers, which means they cannot do numeric calculations.
You can see from the following command.
"""
msa_by_pop.dtypes

Rank                                        int64
MSA                                        object
population_2021_est                         int64
population_2020                             int64
perc_change                                object
Encompassing_combined_statistical_area     object
year_change                               float64
dtype: object

Now using your new `year_change` column, created another column called `population_2021_estimate` and estimate the 2021 population counts using the formula above. 

In [9]:
## Insert your code here
msa_by_pop['population_2021_estimate'] = (msa_by_pop['population_2020'] * (1 + msa_by_pop['year_change'])).astype(int)
### Population should be integer.
msa_by_pop


Rank                                           MSA  population_2021_est  \
0       1  New York-Newark-Jersey City, NY-NJ-CT-PA MSA             19768458   
1       2        Los Angeles-Long Beach-Anaheim, CA MSA             12997353   
2       3        Chicago-Naperville-Elgin, IL-IN-WI MSA              9509934   
3       4           Dallas-Fort Worth-Arlington, TX MSA              7759615   
4       5      Houston-The Woodlands-Sugar Land, TX MSA              7206841   
..    ...                                           ...                  ...   
379   380                              Danville, IL MSA                73095   
380   381                           Lewiston, ID-WA MSA                64851   
381   382                           Walla Walla, WA MSA                62682   
382   383                                  Enid, OK MSA                61926   
383   384                           Carson City, NV MSA                58993   

     population_2020 perc_change  Encompassing_combined_statistical_area  \
0           20140470      −1.85%        New York-Newark, NY-NJ-CT-PA CSA   
1           13200998      −1.54%          Los Angeles-Long Beach, CA CSA   
2            9618502      −1.13%        Chicago-Naperville, IL-IN-WI CSA   
3            7637387      +1.60%            Dallas-Fort Worth, TX-OK CSA   
4            7122240      +1.19%           Houston-The Woodlands, TX CSA   
..               ...         ...                                     ...   
379            74188      −1.47%                                     NaN   
380            64375      +0.74%                                     NaN   
381            62584      +0.16%  Kennewick-Richland-Walla Walla, WA CSA   
382            62846      −1.46%                                     NaN   
383            58639      +0.60%        Reno-Carson City-Fernley, NV CSA   

     year_change  population_2021_estimate  
0      -0.018471                  19768458  
1      -0.015426                  12997353  
2      -0.011287                   9509934  
3       0.016004                   7759615  
4       0.011878                   7206841  
..           ...                       ...  
379    -0.014733                     73095  
380     0.007394                     64851  
381     0.001566                     62682  
382    -0.014639                     61926  
383     0.006037                     58993  

[384 rows x 8 columns]

## 4.2 Exercise 2 (5 points)
In this exercise, we are going to explore urban population change over the previous decade. 

Here, we are going to read the Excel file on "Annual Estimates of the Resident Population: April 1, 2010 to July 1, 2019" directly from the [U.S. Census Bureau's website](https://www.census.gov/data/tables/time-series/demo/popest/2010s-total-metro-and-micro-statistical-areas.html).

Below, I'm going to do some data cleaning and re-formatting so the table is easier to work with.

Each column with the year is the Census estimated MSA population for that year.

In [10]:
# .read_excel() reads an Excel spreadsheet from a local file or from a URL. 
# Most .read_*() methods can read from both local files and URLs.
# The default sheet is the first one, but you can specify a different sheet by name or index.

# skiprows=3 skips the first 3 rows, header=0 uses the first row as column names
msa_pop_2010_2019 = pd.read_excel('https://www2.census.gov/programs-surveys/popest/tables/2010-2019/metro/totals/cbsa-met-est2019-annres.xlsx',
                                skiprows=3,header=0) 

# Because this is an Excel file with multiple header-type rows, 
# we need to skip the first 3 rows
msa_pop_2010_2019 = msa_pop_2010_2019.iloc[2:]

# I'm going to rename the columns to make them easier to work with
msa_pop_2010_2019.columns = ['MSA','apr_1_2010','Estimates Base','2010','2011','2012','2013','2014','2015','2016','2017','2018','2019']

# And I'm selecting only the columns we need for this exercise
msa_pop_2010_2019 = msa_pop_2010_2019[['MSA','2010','2011','2012','2013','2014','2015','2016','2017','2018','2019']]

# And I'm going to remove the rows that are not Metropolitan Statistical Areas 
# by filtering on the string 'Metro Area'
msa_pop_2010_2019 = msa_pop_2010_2019[msa_pop_2010_2019['MSA'].str.contains('Metro Area')]

# There's strangely a period at the beginning of each MSA name, so I'm going to remove it
msa_pop_2010_2019['MSA'] = msa_pop_2010_2019['MSA'].str.replace('.','')

# And I'm going to remove the word 'Metro Area' from the end of each MSA name
msa_pop_2010_2019['MSA'] = msa_pop_2010_2019['MSA'].str.replace(' Metro Area','')

/var/folders/_7/lymm6gzn0cd9kkph02m8lhsc0000gn/T/ipykernel_11842/1369159727.py:24: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  msa_pop_2010_2019['MSA'] = msa_pop_2010_2019['MSA'].str.replace('.','')


Let's take a look at the first 5 rows of the data.

In [11]:
## Insert your code here
msa_pop_2010_2019.head(5)

MSA      2010      2011      2012      2013  \
2                  Abilene, TX  165585.0  166634.0  167442.0  167473.0   
3                    Akron, OH  703031.0  703200.0  702109.0  703621.0   
4                   Albany, GA  154145.0  154545.0  153976.0  152667.0   
5           Albany-Lebanon, OR  116891.0  118164.0  118273.0  118405.0   
6  Albany-Schenectady-Troy, NY  871082.0  872778.0  874698.0  877065.0   

       2014      2015      2016      2017      2018      2019  
2  168342.0  169688.0  170017.0  170429.0  171150.0  172060.0  
3  704908.0  704382.0  703524.0  703987.0  703855.0  703479.0  
4  151949.0  150387.0  149137.0  148090.0  147840.0  146726.0  
5  119042.0  120236.0  122769.0  125035.0  127451.0  129749.0  
6  878113.0  879085.0  879792.0  882158.0  882263.0  880381.0

Now, create a new column that represents the percentage change in population between 2010 and 2019 and call this column `perc_pop_change`. 

In [12]:
## Insert your code here

msa_pop_2010_2019['perc_pop_change'] = msa_pop_2010_2019['2019'] / msa_pop_2010_2019['2010'] - 1
msa_pop_2010_2019

MSA       2010       2011       2012       2013  \
2                    Abilene, TX   165585.0   166634.0   167442.0   167473.0   
3                      Akron, OH   703031.0   703200.0   702109.0   703621.0   
4                     Albany, GA   154145.0   154545.0   153976.0   152667.0   
5             Albany-Lebanon, OR   116891.0   118164.0   118273.0   118405.0   
6    Albany-Schenectady-Troy, NY   871082.0   872778.0   874698.0   877065.0   
..                           ...        ...        ...        ...        ...   
423                 Mayagüez, PR   115929.0   113915.0   111845.0   109938.0   
424                    Ponce, PR   261924.0   257719.0   253345.0   249504.0   
425               San Germán, PR   137630.0   136415.0   135185.0   133955.0   
426  San Juan-Bayamón-Caguas, PR  2347932.0  2322216.0  2295419.0  2269908.0   
427                    Yauco, PR   106937.0   104954.0   102970.0   101054.0   

          2014       2015       2016       2017       2018       2019  \
2     168342.0   169688.0   170017.0   170429.0   171150.0   172060.0   
3     704908.0   704382.0   703524.0   703987.0   703855.0   703479.0   
4     151949.0   150387.0   149137.0   148090.0   147840.0   146726.0   
5     119042.0   120236.0   122769.0   125035.0   127451.0   129749.0   
6     878113.0   879085.0   879792.0   882158.0   882263.0   880381.0   
..         ...        ...        ...        ...        ...        ...   
423   107458.0   105048.0   102580.0    99742.0    95378.0    94975.0   
424   244760.0   239275.0   233160.0   226541.0   216312.0   215295.0   
425   132195.0   130238.0   128290.0   125549.0   121080.0   121464.0   
426  2233776.0  2195941.0  2155066.0  2104671.0  2022139.0  2023227.0   
427    98695.0    96258.0    93747.0    90758.0    86497.0    85830.0   

     perc_pop_change  
2           0.039104  
3           0.000637  
4          -0.048130  
5           0.110000  
6           0.010675  
..               ...  
423        -0.180749  
424        -0.178025  
425        -0.117460  
426        -0.138294  
427        -0.197378  

[392 rows x 12 columns]

Print the average population change for the 10 **largest** MSAs by population in 2010. 

In [13]:
# Insert your code here
change_top_10 = msa_pop_2010_2019.sort_values('2010',ascending = False).head(10)['perc_pop_change'].mean()

# The print function is used to print a string to the screen
print('The average population change for the 10 largest MSAs between 2010 and 2019 is', 
        round(change_top_10*100,4),'%')

The average population change for the 10 largest MSAs between 2010 and 2019 is 8.5142 %


Now print the average population change for the 10 **smallest** MSAs by population in 2010. 

In [14]:
# Insert your code here
change_bottom_10 = msa_pop_2010_2019.sort_values('2010',ascending = True).head(10)['perc_pop_change'].mean()

print('The average population change for the 10 smallest MSAs between 2010 and 2019 is', 
    round(change_bottom_10*100,4),'%')

The average population change for the 10 smallest MSAs between 2010 and 2019 is 2.4318 %
